### PR TITLE
Feat: (gatsby-cli) Add a CLI command for listing plugins

### PR DESCRIPTION
**Pull request by laurieontech**
The command makes use of existing gatsby recipe provider functionality.

[ch17790]

### CODE REVIEW

File: integration-tests/gatsby-cli/gatsby-sites/gatsby-plugin/package.json [hunk 1]
```diff
@@ -0,0 +1,39 @@
+{
+  "name": "gatsby-starter-default-develop",
```
Comment (laurieontech) [inline integration-tests/gatsby-cli/gatsby-sites/gatsby-plugin/package.json:2]:
```suggestion
  "name": "gatsby-starter-default-plugin",
```

File: packages/gatsby-recipes/src/providers/gatsby/plugin.js [hunk 2]
```diff
@@ -511,11 +511,13 @@ export { create, create as update, read, destroy }
 
 export const config = {}
 
-export const all = async ({ root }) => {
+export const all = async ({ root }, processPlugins = true) => {
   const configSrc = await readConfigFile(root)
   const plugins = getPluginsFromConfig(configSrc)
 
-  return Promise.all(plugins.map(({ name }) => read({ root }, name)))
+  return Promise.all(
```
Comment (gillkyle) [inline packages/gatsby-recipes/src/providers/gatsby/plugin.js:518]: I don't think this works with local plugins. I think it's a result of reading the config files and get the plugins, but I keep getting big errors like this:

```
Error: Cannot find module 'gatsby-plugin-test/package.json'
Require stack:
- /Users/kyle/Github/gatsby/starters/hello-world/node_modules/gatsby-cli/node_modules/gatsby-recipes/dist/entry-74a70aa0.js
- /Users/kyle/Github/gatsby/starters/hello-world/node_modules/gatsby-cli/node_modules/gatsby-recipes/dist/index.js
- /Users/kyle/Github/gatsby/starters/hello-world/node_modules/gatsby-cli/lib/recipes.js
- /Users/kyle/Github/gatsby/starters/hello-world/node_modules/gatsby-cli/lib/create-cli.js
- /Users/kyle/Github/gatsby/starters/hello-world/node_modules/gatsby-cli/lib/index.js
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:966:15)
    at Function.resolve (internal/modules/cjs/helpers.js:78:19)
    at listShadowableFilesForTheme (/Users/kyle/Github/gatsby/starters/hello-world/node_modules/gatsby-cli/node_modules/gatsby-recipes/src/providers/gatsby/plugin.js:28:13)
    at read$4 (/Users/kyle/Github/gatsby/starters/hello-world/node_modules/gatsby-cli/node_modules/gatsby-recipes/src/providers/gatsby/plugin.js:275:50)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
    at async Promise.all (index 0)
    at _default (/Users/kyle/Github/gatsby/starters/hello-world/node_modules/gatsby-cli/lib/handlers/plugin.js:39:25)
    at Object.handler (/Users/kyle/Github/gatsby/starters/hello-world/node_modules/gatsby-cli/lib/create-cli.js:358:7) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/Users/kyle/Github/gatsby/starters/hello-world/node_modules/gatsby-cli/node_modules/gatsby-recipes/dist/entry-74a70aa0.js',
    '/Users/kyle/Github/gatsby/starters/hello-world/node_modules/gatsby-cli/node_modules/gatsby-recipes/dist/index.js',
    '/Users/kyle/Github/gatsby/starters/hello-world/node_modules/gatsby-cli/lib/recipes.js',
    '/Users/kyle/Github/gatsby/starters/hello-world/node_modules/gatsby-cli/lib/create-cli.js',
    '/Users/kyle/Github/gatsby/starters/hello-world/node_modules/gatsby-cli/lib/index.js'
  ]
}

 ERROR 

gatsby plugin <cmd> [plugins...]

Useful commands relating to Gatsby plugins

Positionals:
  cmd      Valid commands include `docs`, `ls`.
                      [string] [required] [choices: "docs", "ls"]
  plugins  The plugin names                 [array] [default: []]

Options:
  --verbose                Turn on verbose output
                                       [boolean] [default: false]
  --no-color, --no-colors  Turn off the color in output
                                       [boolean] [default: false]
  --json                   Turn on the JSON logger
                                       [boolean] [default: false]
  -h, --help               Show help                    [boolean]
  -v, --version            Show the version of the Gatsby CLI and
                           the Gatsby package in the current
                           project                      [boolean]


 ERROR 

There was an error


 ERROR 

Error: Cannot find module 'gatsby-plugin-test/package.json'
Require stack:
- /Users/kyle/Github/gatsby/starters/hello-world/node_modules/gatsby-cli/node_modules/gatsby-recipes/dist/entry-74a70aa0.js
- /Users/kyle/Github/gatsby/starters/hello-world/node_modules/gatsby-cli/node_modules/gatsby-recipes/dist/index.js
- /Users/kyle/Github/gatsby/starters/hello-world/node_modules/gatsby-cli/lib/recipes.js
- /Users/kyle/Github/gatsby/starters/hello-world/node_modules/gatsby-cli/lib/create-cli.js
- /Users/kyle/Github/gatsby/starters/hello-world/node_modules/gatsby-cli/lib/index.js
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:966:15)
    at Function.resolve (internal/modules/cjs/helpers.js:78:19)
    at listShadowableFilesForTheme (/Users/kyle/Github/gatsby/starters/hello-world/node_modules/gatsby-cli/node_modules/gatsby-recipes/src/providers/gatsby/plugin.js:28:13)
    at read$4 (/Users/kyle/Github/gatsby/starters/hello-world/node_modules/gatsby-cli/node_modules/gatsby-recipes/src/providers/gatsby/plugin.js:275:50)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
    at async Promise.all (index 0)
    at _default (/Users/kyle/Github/gatsby/starters/hello-world/node_modules/gatsby-cli/lib/handlers/plugin.js:39:25)
    at Object.handler (/Users/kyle/Github/gatsby/starters/hello-world/node_modules/gatsby-cli/lib/create-cli.js:358:7) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/Users/kyle/Github/gatsby/starters/hello-world/node_modules/gatsby-cli/node_modules/gatsby-recipes/dist/entry-74a70aa0.js',
    '/Users/kyle/Github/gatsby/starters/hello-world/node_modules/gatsby-cli/node_modules/gatsby-recipes/dist/index.js',
    '/Users/kyle/Github/gatsby/starters/hello-world/node_modules/gatsby-cli/lib/recipes.js',
    '/Users/kyle/Github/gatsby/starters/hello-world/node_modules/gatsby-cli/lib/create-cli.js',
    '/Users/kyle/Github/gatsby/starters/hello-world/node_modules/gatsby-cli/lib/index.js'
  ]
}
```

I tried adding a local plugin with an empty package.json, and a filled out package.json, both gave this error. I got it outputting info from gatsby-source-filesystem. It's possible I'm doing something wrong testing this 🤔

**Review by gillkyle — COMMENTED** [review]
I'm not sure this plays nicely with local plugins (though maybe I'm the only one who uses those haha), but I've also had some trouble testing this reliably so I'd be happily disputed on that point.

File: packages/gatsby-recipes/src/providers/gatsby/plugin.js (hunk 2 above)
Comment (gillkyle) [reply packages/gatsby-recipes/src/providers/gatsby/plugin.js:518]: (we got this working so I'ma resolve the conversation, it seemed to be a problem with my setup somewhere)

**Review by mxstbr — APPROVED** [review]
Let's do it. No clue why I didn't approve this earlier, sorry about that!

File: packages/gatsby-cli/src/create-cli.ts [hunk 3]
```diff
@@ -400,15 +401,21 @@ function buildLocalCommands(cli: yargs.Argv, isLocalSite: boolean): void {
     builder: yargs =>
       yargs
         .positional(`cmd`, {
-          choices: [`docs`],
-          describe: "Valid commands include `docs`.",
+          choices: [`docs`, `ls`],
```
Comment (ascorbic) [inline packages/gatsby-cli/src/create-cli.ts:404]: Not needed now, but I wonder if we should add `list` as an alias for `ls`. Windows people might not know what `ls` means

**Review by ascorbic — APPROVED** [review]
:shipit:

File: packages/gatsby-cli/src/create-cli.ts (hunk 3 above)
Comment (laurieontech) [reply packages/gatsby-cli/src/create-cli.ts:404]: Fair enough and easy to do. Let’s get this in and we can add it in the next pass.

**Pull request merged**